beacon-chain/sync/block_batcher.go

```diff
@@ -0,0 +1,210 @@
+package sync
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	libp2pcore "github.com/libp2p/go-libp2p/core"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/db/filters"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+// blockRangeBatcher encapsulates the logic for splitting up a block range request into fixed-size batches of
+// blocks that are retrieved from the database, ensured to be canonical, sequential and unique.
+// If a non-nil value for ticker is set, it will be used to pause between batches lookups, as a rate-limiter.
+type blockRangeBatcher struct {
+	start   primitives.Slot
+	end     primitives.Slot
+	size    uint64
+	db      db.NoHeadAccessDatabase
+	limiter *limiter
+	ticker  *time.Ticker
+
+	cf      *canonicalFilter
+	current *blockBatch
+}
+
+func newBlockRangeBatcher(rp rangeParams, bdb db.NoHeadAccessDatabase, limiter *limiter, canonical canonicalChecker, ticker *time.Ticker) *blockRangeBatcher {
```

Comment:
how about functional opts for the ticker instead of having to pass in nil?

Reply:
I thought about that but this will probably only ever have 2 users (BlocksByRange and BlobSidecarsByRange), both of which will set the timeout, so I thought the extra boilerplate almost not worth it. The only reason I thought it would be useful to not set the timer was for tests, so maybe I should just do a nil check and enforce it, since tests can directly populate the struct?